### PR TITLE
feat: Build Docker images

### PR DESCRIPTION
**Pull request by kuisathaverat**
## What does this PR do?

It adds a new stage that will build and push it to the`docker.elastic.co` Docker registry.

## Why is it important?

Every commit that passes the build and test stages will build a docker image, 
then it will push the image using the commit SH1 as a tag (`docker.elastic.co/package-registry/package-registry:f999b7a84d977cd19a379f0cec802aa1ef7ca379`), 
finally, we will push another tag using the git branch or tag name (`docker.elastic.co/package-registry/package-registry:master`,
`docker.elastic.co/package-registry/package-registry:pr-111`,
`docker.elastic.co/package-registry/package-registry:v0.2.0`)

## Related issues
Closes https://github.com/elastic/package-registry/issues/205

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -158,6 +158,16 @@ docker build --rm -t integrations_registry:latest .
 docker run -i -t -p 8080:8080 $(docker images -q integrations_registry:latest)
 ```
 
+#### Docker images published
+
+We publish a Docker image with each successful build commit on branches, tags, or PR.
+For each commit we have two docker image tags, one wit the commit as tag
```
Comment (ruflin) [inline README.md]:
```suggestion
For each commit we have two docker image tags, one with the commit as tag
```

File: README.md [hunk 2]
```diff
@@ -158,6 +158,16 @@ docker build --rm -t integrations_registry:latest .
...
+`docker.elastic.co/package-registry/package-registry:f999b7a84d977cd19a379f0cec802aa1ef7ca379`
+An another Docker tag with the git branch or tag name
+`docker.elastic.co/package-registry/package-registry:master`,
+`docker.elastic.co/package-registry/package-registry:pr-111`,
+`docker.elastic.co/package-registry/package-registry:v0.2.0`
```
Comment (ruflin) [inline README.md]: Will a tag automatically pushed as soon as one is created?

For branches, I assume `master` is overwritten after each merge (which is expected). As most PR's don't have a branch in this repo but are PR's from a fork, I assume no docker image it built for it (expected):

File: Jenkinsfile [hunk 3]
```diff
@@ -90,6 +85,31 @@ pipeline {
...
+          dockerLogin(secret: "${env.DOCKER_REGISTRY_SECRET}",
+            registry: "${env.DOCKER_REGISTRY}")
+          sh(label: 'Build Docker image',
+            script: "docker build -t ${env.DOCKER_IMG_TAG} .")
+          sh(label: 'Push Docker image',
+            script: "docker push ${env.DOCKER_IMG_TAG}")
+          sh(label: 'Re-tag Docker image',
+            script: "docker tag ${env.DOCKER_IMG_TAG} ${env.DOCKER_IMG_TAG_BRANCH}")
+            sh(label: 'Push Docker image',
+              script: "docker push ${env.DOCKER_IMG_TAG_BRANCH}")
```
Comment (v1v) [inline Jenkinsfile]: does the push requires a --force flag when the tag already exists?

File: Jenkinsfile [hunk 4]
```diff
@@ -90,6 +85,31 @@ pipeline {
...
+        dir("${BASE_DIR}"){
+          dockerLogin(secret: "${env.DOCKER_REGISTRY_SECRET}",
+            registry: "${env.DOCKER_REGISTRY}")
+          sh(label: 'Build Docker image',
+            script: "docker build -t ${env.DOCKER_IMG_TAG} .")
+          sh(label: 'Push Docker image',
+            script: "docker push ${env.DOCKER_IMG_TAG}")
+          sh(label: 'Re-tag Docker image',
+            script: "docker tag ${env.DOCKER_IMG_TAG} ${env.DOCKER_IMG_TAG_BRANCH}")
+            sh(label: 'Push Docker image',
```
Comment (v1v) [inline Jenkinsfile]:
```suggestion
            sh(label: 'Push Docker image name',
```

File: Jenkinsfile [hunk 5]
```diff
@@ -90,6 +85,31 @@ pipeline {
...
+      }
+      steps {
+        deleteDir()
+        unstash 'source'
+        dir("${BASE_DIR}"){
+          dockerLogin(secret: "${env.DOCKER_REGISTRY_SECRET}",
+            registry: "${env.DOCKER_REGISTRY}")
+          sh(label: 'Build Docker image',
+            script: "docker build -t ${env.DOCKER_IMG_TAG} .")
+          sh(label: 'Push Docker image',
```
Comment (v1v) [inline Jenkinsfile]:
```suggestion
          sh(label: 'Push Docker image sha',
```

File: README.md (hunk 2 above)
Comment (kuisathaverat) [reply README.md]: it does not matter, every PR has a remote ref on GitHub. Jenkins uses the name of the remote ref as branch name, something like PR-1, PR-2,...PR-100

**Comment by kuisathaverat** [comment]
>I'm somehow concerned that we end up with too many docker images as each commit into master will have one, at the same time it is exactly what we needed.

This is the intention, we could test any commit that passes the CI without building the image we have it. Because of how Docker images are stores we only store the changes between images to it is not a concern the disk space needed.

File: Jenkinsfile (hunk 3 above)
Comment (kuisathaverat) [reply Jenkinsfile]: no, it is not needed, we push without `--force` here and it is working https://github.com/elastic/apm-pipeline-library/blob/master/.ci/buildDockerImages.groovy#L340

File: README.md (hunk 2 above)
Comment (ruflin) [reply README.md]: Yes, but do we really need a Docker image for each PR? Having it only for things that were merged into master and the branches in `elastic/package-registry` should be enough.

**Comment by ruflin** [comment]
@kuisathaverat Seems like something goes wrong in the CI when building the docker image.

**Comment by kuisathaverat** [comment]
@ruflin The credentials are not valid, I have reopened https://github.com/elastic/infra/issues/17661 because of that.

**Comment by kuisathaverat** [comment]
@ruflin everything is working now, merging .....

**Pull request merged**